components/pkg-export-docker/src/chmod.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2018 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Encapsulates code to perform the equivalent of `chmod -R g=u` on a
+//! directory.
+//!
+//! This is needed in order to ensure that the `/hab` directory in a
+//! BuildRoot is constructed such that the root group has write
+//! permissions, allowing containerized Habitat services to run as
+//! non-root users, which is necessary for running in platforms like
+//! OpenShift and various HPC situations.
+//!
+//! Because Docker does not yet provide the ability within a
+//! Dockerfile to perform an `ADD` or `COPY` command while also
+//! setting permissions on those files, we must set the permissions on
+//! the files _before_ they are pulled into the container. We could
+//! add a `RUN chmod -R g=u /hab` in the Dockerfile, but that
+//! essentially doubles the size of any container, due to how Docker's
+//! layered filesystem works. The approach here takes more code, but
+//! results in a better experience for end-users.
+//!
+//! Also note that this is currently a no-op on Windows, since
+//! permissions work differently over there, and we don't have
+//! non-root Habitat supervisors there either.
+
+use std::fs;
+#[cfg(windows)]
+use std::fs::Permissions;
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+
+use error::Result;
+#[cfg(unix)]
+use hcore::util::perm;
+
+/// Perform the equivalent of `chmod -R g=u path`.
+pub fn recursive_g_equal_u<P>(path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let path = path.as_ref().to_path_buf();
+    let metadata = fs::symlink_metadata(&path)?;
+    let permissions = metadata.permissions();
+    let filetype = metadata.file_type();
+
+    if filetype.is_symlink() {
+        // skip
+    } else if filetype.is_dir() {
+        set_g_equal_u(&path, permissions)?;
+        for entry in fs::read_dir(&path)? {
+            let entry = entry.unwrap().path();
+            recursive_g_equal_u(entry)?;
```

Comment:
I'm a bit concerned that, unchecked we could keep recursing until the stack fills and Rust has to unwind. Mind you that could be a fair number of stack frames, but I'm wondering if it's worth having a max number check? Then again, what is this value?

Reply:
I suppose we could add a limit... it'd be the longest file path under `/hab`, though, which should be fine in all but the most pathological case. We're already not following symlinks, so we at least don't have to worry about cycles.

I could pick a ridiculous number, just to cap things concretely... 100?

Reply:
@fnichol Actually, it looks like Rust uses recursion in the implementation of some of its file-manipulation functions (as with the underlying implementation of [DirBuilder::new().recursive(true).create(...)](https://doc.rust-lang.org/1.23.0/src/std/fs.rs.html#1846))... so we're probably fine with the current implementation.